pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -0,0 +1,292 @@
+		ep, ok := curr.(*discoveryv1alpha1.EndpointSlice)
+		if !ok {
+			tombstone, ok := curr.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				log.Errorf("Couldn't get object from tombstone %#v", curr)
+				return nil
+			}
+			ep, ok = tombstone.Obj.(*discoveryv1alpha1.EndpointSlice)
+			if !ok {
+				log.Errorf("Tombstone contained an object that is not an endpoints slice %#v", curr)
+				return nil
+			}
+		}
+
+		e.updateEDSSlice(c, ep, event)
+
+		return nil
+	})
+}
+
+func (e endpointSliceController) GetEndpointServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance {
+	endpointsForPodInSameNS := make([]*model.ServiceInstance, 0)
+	endpointsForPodInDifferentNS := make([]*model.ServiceInstance, 0)
+
+	for _, item := range e.cache.informer.GetStore().List() {
+		slice := *item.(*discoveryv1alpha1.EndpointSlice)
+		endpoints := &endpointsForPodInSameNS
+		if slice.Namespace != proxyNamespace {
+			endpoints = &endpointsForPodInDifferentNS
+		}
+
+		*endpoints = append(*endpoints, c.getProxyServiceInstancesByEndpointSlice(slice, proxy)...)
+	}
+
+	// Put the endpointsForPodInSameNS in front of endpointsForPodInDifferentNS so that Pilot will
+	// first use endpoints from endpointsForPodInSameNS. This makes sure if there are two endpoints
+	// referring to the same IP/port, the one in endpointsForPodInSameNS will be used. (The other one
+	// in endpointsForPodInDifferentNS will thus be rejected by Pilot).
+	return append(endpointsForPodInSameNS, endpointsForPodInDifferentNS...)
+}
+
+func (e *endpointSliceController) InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
+	labelsList labels.Collection) ([]*model.ServiceInstance, error) {
+	esLabelSelector := klabels.Set(map[string]string{discoveryv1alpha1.LabelServiceName: svc.Attributes.Name}).AsSelectorPreValidated()
+	var slices []*discoveryv1alpha1.EndpointSlice
+	err := cache.ListAllByNamespace(e.cache.informer.GetIndexer(), svc.Attributes.Namespace, esLabelSelector, func(i interface{}) {
+		slices = append(slices, i.(*discoveryv1alpha1.EndpointSlice))
+	})
+	if err != nil {
+		log.Infof("get endpoints(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return nil, nil
+	}
+	if len(slices) == 0 {
+		return nil, nil
+	}
+
+	// Locate all ports in the actual service
+	svcPortEntry, exists := svc.Ports.GetByPort(reqSvcPort)
+	if !exists {
+		return nil, nil
+	}
+
+	var out []*model.ServiceInstance
+	for _, slice := range slices {
+		for _, e := range slice.Endpoints {
+			for _, a := range e.Addresses {
+				var podLabels labels.Instance
+				pod := c.pods.getPodByIP(a)
+				if pod != nil {
+					podLabels = configKube.ConvertLabels(pod.ObjectMeta)
+				}
+
+				// check that one of the input labels is a subset of the labels
+				if !labelsList.HasSubsetOf(podLabels) {
+					continue
+				}
+
+				az, sa, uid := "", "", ""
+				if pod != nil {
+					az = c.GetPodLocality(pod)
```

Comment:
should make use of discovery.Endpoint.Topology

Reply:
thanks, I switched this